### ISSUE TITLE
Check "model consistency" for users during model construction?

### ISSUE THREAD

**Issue by ali-ramadhan** [CliMA/Oceananigans.jl#1453]
I was thinking it might be cool/useful to inform users if their model is physically/numerically consistent.

One example of this is to check that the boundary conditions being imposed match the grid topology (#890).

But I think there are many more. Some examples:
* Using `BetaPlane` with periodic boundary conditions.
* Using `WENO5` on a curvilinear grid.
* Using a free surface when z is `Periodic`.
* Using an `FFTBasedPressureSolver` on a curvilinear grid.
* ...

Some errors are more likely to be made by users than others of course. But I don't think we can expect users to just know that e.g. `WENO5` doesn't work with `VerticallyStretchedRectilinearGrid` yet.

To save ourselves and users some potential future headaches, it seems like we could create a function like `check_consistency(model)` that would perform these checks and print a warning for each inconsistency detected.

Unfortunately this could become a pretty ugly monolithic function that's hard to modularize since we're checking the consistency of many different structs acting together.

**Comment by ali-ramadhan** [comment]
Yeah that would definitely help with modularization. They might need to take in the whole model, e.g. `validate_coriolis(model)`, `validate_advection(model)`, etc. since it seems hard to anticipate all the inconsistent configurations.

I remember you raised the concern of slowdowns in the case of running tons of small/column models. So maybe a kwarg like `check_consistency=true` would be good to add with this feature.

**Comment by glwagner** [comment]
That could work --- though I'd point out that listing the model components that can be inconsistent with one another might be more explicit.

I would advocate for a "blacklisting" method whereby blacklisted combinations of components are listed local to the implementation of the component. For example, in `beta_plane.jl` we could add

```julia
validate_coriolis(::BetaPlane, ::AbstractGrid{FT, TX, <:Periodic}) where {FT, TX} = # error
```

we also need, in `hydrostatic_spherical_coriolis.jl`,

```julia
validate_coriolis(::HydrostaticSphericalCoriolis, ::AbstractRectilinearGrid) = # error
```

I think a warning is better than an error for `WENO5` on a curvilinear grid. That'd be a fun scheme to experiment with and we don't know "how wrong" it is. If the curvilinear grid is almost regular, we have reason to think it might work. In cases like these where we aren't completely, rock-solid positive that a certain model configuration is invalid I think we should also err conservative and allow such models to be constructed. 

We do need to strike a balance between being helicopter parents to users, and catching common mistakes. I don't think we should get in the business of curating science --- we should just provide useful software.

Specialty applications can just avoid the kwarg-based constructor so I'm not sure we should worry much about that.

**Comment by ali-ramadhan** [comment]
> We do need to strike a balance between being helicopter parents to users, and catching common mistakes. I don't think we should get in the business of curating science --- we should just provide useful software.

Yes to be clear I should have said that I'm just advocating for printing warnings, not throwing errors. Users should be able to run whatever they want. I would say that these warnings make the software more useful.

**Comment by ali-ramadhan** [comment]
Another idea: check that the grid architecture and model architecture match to avoid issues like the test failure in #1467.

**Comment by glwagner** [comment]
Should users only have to specify `architecture` in one place?

**Comment by ali-ramadhan** [comment]
Yes that would be ideal!

Actually I guess right now it might be tricky since you have to create the grid first for stretched and curvilinear grids. But for multi-architectures, you need to pass the grid as the grid topology could change the rank connectivity. So you end up having to specify `CPU()` for the grid, then pass it to the grid, then specify `MultiCPU(...)` for the model.

Would be great if we can specify the architecture in one place, but might require some refactoring that we'd have to think more about.

**Comment by glwagner** [comment]
We could change the API so that `architecture` is specified with `grid` always?

**Comment by ali-ramadhan** [comment]
Ah like adding an `architecture` kwarg that does nothing for `RegularRectilinearGrid`? I think that would be a good idea.

I would generally also support going with less defaults, so making `architecture` an explicit always-required kwarg.
